draft-ietf-tls-esni.md

```diff
@@ -710,26 +707,85 @@ if a client proposes ALPN values in ClientHelloInner, the server-selected value
+ECH. Otherwise, if it has a length other than 8, the client aborts the handshake
+with a "decode_error" alert. Otherwise, the client computes
+`hrr_accept_confirmation` as described in {{backend-server-hrr}}. If this value
+matches the extension payload, the server has accept ECH. Otherwise, it has
+rejected ECH.
+
+[[OPEN ISSUE: Depending on what we do for issue#450, it may be appropriate to
+change the client behavior if the HRR extension is present but with the wrong
+value.]]
 
-If the server used ClientHelloInner, the client proceeds with the connection as
-usual, authenticating the connection for the true server name.
+If the server accepts ECH, the client handshakes with ClientHelloInner as
+described in {{accepted-ech}}. Otherwise, the client handshakes with
+ClientHelloOuter as described in {{rejected-ech}}.
 
-#### Rejected ECH {#rejected-ech}
+### Handshaking with ClientHelloInner {#accepted-ech}
 
-If the server used ClientHelloOuter, the client proceeds with the handshake,
+If the server accepts ECH, the client proceeds with the connection as in
+{{RFC8446}}, with the following modifications:
+
+The client behaves as if it had sent ClientHelloInner as the ClientHello. That
+is, it evaluates the handshake using the ClientHelloInner's preferences, and,
+when computing the transcript hash ({{Section 4.4.1 of RFC8446}}), it uses
+ClientHelloInner as the first ClientHello.
+
+If the server responds with a HelloRetryRequest, the client computes the updated
+ClientHello message as follows:
+
+1. It computes a second ClientHelloInner based on the first ClientHelloInner, as
+   in {{Section 4.1.4 of RFC8446}}. The ClientHelloInner's
+   "encrypted_client_hello" extension is left unmodified.
+
+1. It constructs EncodedClientHelloInner as described in {{encoding-inner}}.
+
+1. It constructs a second partial ClientHelloOuterAAD message. This message MUST
+   be syntactically valid. The extensions MAY be copied from the original
+   ClientHelloOuter unmodified, or omitted. If not sensitive, the client MAY
+   copy updated extensions from the second ClientHelloInner for compression.
+
+1. It encrypts EncodedClientHelloInner as described in
+   {{encrypting-clienthello}}, using the second partial ClientHelloOuterAAD, to
+   obtain a second ClientHelloOuter. It reuses the original HPKE encryption
+   context computed in {{real-ech}} and uses the empty string for `enc`.
+
+   The HPKE context maintains a sequence number, so this operation internally
+   uses a fresh nonce for each AEAD operation. Reusing the HPKE context avoids
+   an attack described in {{flow-hrr-hijack}}.
+
+Note this procedure does not apply {{Section 4.1.4 of RFC8446}} to the
+ClientHelloOuter, only ClientHelloInner. The server generates HelloRetryRequest
+from ClientHelloInner. If the ClientHelloOuter and ClientHelloInner expressed
+different preferences, applying the standard procedure to ClientHelloOuter may
+not succeed.
```

Comment:
```suggestion
```
This is covered by the first step (1), right?

Reply:
Yeah, I guess there's no real need to restate it. Figured it wasn't obvious that the HelloRetryRequest might not match.